### DISCUSSION TITLE
Travis continuous integration is no longer an option; need new location

### DISCUSSION THREAD

**Discussion by parrt** [antlr/antlr4#3031] [Ideas]
Howdy! @ericvergnaud  has graciously agreed to lead a discussion to figure out how to get continuous integration up and running again. The good news is that we at least have appveyor still working. Further, I just figured out how to use github actions... fairly straightforward actually. Here's an example where I run a bunch of Python unit tests after installing a bunch of libraries:

https://github.com/parrt/tensor-sensor/blob/master/.github/workflows/test.yml

And you get the results that look like this page:

https://github.com/parrt/tensor-sensor/actions/runs/467542701

**Comment by parrt** [comment]
CircleCI Comes highly recommended to me and they have an [open source option](https://circleci.com/open-source/), though I'm not sure how much `400,000 credits per month for open source builds` would cover.

**Comment by ericvergnaud** [comment]
Hi,

I've put together a tentative comparison of free offerings.
![image](https://user-images.githubusercontent.com/3332132/103983906-1fdc4780-51c1-11eb-93ba-e1db5c43d0b4.png)

Only AppVeyor offers unlimited builds, but each of them would take 5.3 hours.

Circle CI are very OSS friendly, but they make it clear that their generosity does not span beyond Linux.

So it seems the only option to recover our unlimited builds capability across all required environments would be to rely on self-hosted runners, which would also maximize parallel jobs and reduce the overall duration down to 30mns (duration of our longest job).

**Reply by parrt** [reply]
Interesting.   It might cost us a few hundred dollars per year on Amazon. Is that what you mean by self hosted runners?  Can you estimate how many builds we can do on circle before we run out? Instead of paying for an Amazon machine, maybe if we paid circle that would be cheaper and easier?

**Comment by ericvergnaud** [comment]
Self hosted runner is my box, your box, anyone's box. Could be Amazon too.
I like the idea because it potentially reduces overall build time if enough members of the community are willing to act as runners.

**Reply by parrt** [reply]
Yeah, a fast box Where we can run in parallel would be great, but we should avoid having multiple boxes because that is multiple points of failure when trying to get a build released.

 I can potentially get a virtual machine here at university, but it would be Linux only; appveyor can do the Windows part I guess but we don't have a Mac solution.

**Comment by ericvergnaud** [comment]
Circle CI we can do 3 free builds excluding Mac.

**Reply by parrt** [reply]
3 free? is that per day or something?  With 400,000 credits per month on circle and 1500 credits per build, seems like that would be 266 builds per month which is pretty good for free. In my reading that correctly? That is only UNIX of course.

**Comment by ericvergnaud** [comment]
Yes the problem is that's only UNIX
The moment you factor in other platforms, limits become very easy to reach.
It's 3 per month.

**Comment by ericvergnaud** [comment]
I haven't identified a workable free solution for Mac.
This is what leads me to recommend the distributed model. If we have 10 contributors willing to provide build time then the failure rate should become very low.
And if we can achieve that for Mac, then I'm sure we can also achieve it for Wintel and Linux.
(I would hate to have 3 distinct CIs, a nightmare to maintain)

**Comment by parrt** [comment]
I see.  so is there an easy way to have circle launch the tests across multiple servers or are we simply running circle software on our own servers that get notified of commit and pull request events? There are virtual Mac solutions out there but they are about $20 a month or something. I'm sure I could get a Mac M1 mini attached to our network though.

**Comment by ericvergnaud** [comment]
Circle does not not support self hosted runners
GitLab, GitHub and AppVeyor all require an agent running on the self hosted runner.
This agent should run in a VDI or a docker VM to provide some level of isolation.

**Reply by parrt** [reply]
Ok, sorry for being dense, is the idea that a number of us would provide build resources around the web and these would be running an agent from github, for example? Then github would contact that agent and display the results for us at github?

**Reply by ericvergnaud** [reply]
Yes, I am currently experiencing bringing a mac vm (running on my mac) as a self hosted runner

**Comment by parrt** [comment]
I have just updated the @antlr/antlr-targets team to include more recent target contributors. Pinging them here for their input.

**Comment by ericvergnaud** [comment]
Hi,

I just spent a man/day trying to get a GitHub self-hosted runner to work on MacOS for Swift, to no end.
It's able to fetch a commit and run all the builds but fails miserably at running tests (it runs 1 test file and then hangs forever).
If I run the test script from the cmd line it works.
So I suspect it has something to do with the GitHub runner getting confused by the child processes being created, a model which is core to our testing architecture.
It might also be related to the GitHub runner never properly tested on MacOSX (as an example there is no predefined action to install maven...)
If someone wants to try it out feel free...
I will be looking at alternatives.

**Reply by parrt** [reply]
Tests work from cmd-line when i set jdk-7 but not from intellij.  It hangs in the annotation preprocessing every time; it appears that even if you say to use Java seven, it insists upon using its own Java for those preprocessing steps, which no longer work with our annotation code.  can't remember exactly, but there was some proxy class type mismatch.

I should point out that this is even during build not running the tests actually. firing up the development environment now to test again...  I believe this is also after I upgraded the development environment because a few months ago it was okay.

During java compilation, I get the following error message:

```
...
Adding @NotNull assertions... [antlr4-runtime-test-annotations]
Parsing java... [antlr4-runtime-test-annotation-processors]
java: User-specified option "-proc:none" is ignored for "antlr4-runtime-test-annotation-processors". This compilation parameter is set automatically according to project settings.
...

java: java.lang.ClassCastException: com.sun.proxy.$Proxy11 cannot be cast to com.sun.tools.javac.processing.JavacProcessingEnvironment
Abnormal build process termination: 
"/Applications/IntelliJ IDEA.app/Contents/jbr/Contents/Home/bin/java" -Xmx700m -Djava.awt.headless=true -Djdt.compiler.useSingleThread=true -Dcompile.parallel=false -Drebuild.on.dependency.change=true -Djava.net.preferIPv4Stack=true -Dio.netty.initialSeedUniquifier=-8343923017917855334 -Dfile.encoding=UTF-8 -Duser.language=en -Duser.country=US -Didea.paths.selector=IntelliJIdea2020.3 "-Didea.home.path=/Applications/IntelliJ IDEA.app/Contents" "-Didea.config.path=/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3" "-Didea.plugins.path=/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins" -Djps.log.dir=/Users/parrt/Library/Logs/JetBrains/IntelliJIdea2020.3/build-log "-Djps.fallback.jdk.home=/Applications/IntelliJ IDEA.app/Contents/jbr/Contents/Home" -Djps.fallback.jdk.version=11.0.9.1 -Dio.netty.noUnsafe=true -Djava.io.tmpdir=/Users/parrt/Library/Caches/JetBrains/IntelliJIdea2020.3/compile-server/antlr4-master_1350a95b/_temp_ -Djps.backward.ref.index.builder=true -Dkotlin.incremental.compilation=true -Dkotlin.incremental.compilation.js=true -Dkotlin.daemon.enabled -Dkotlin.daemon.client.alive.path=\"/var/folders/93/9kzk2ccm8xj8k70059b28jk80000gp/T/kotlin-idea-11282220729510563313-is-running\" -classpath "/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/jps-launcher.jar" org.jetbrains.jps.cmdline.Launcher "/Applications/IntelliJ IDEA.app/Contents/lib/netty-common-4.1.52.Final.jar:/Applications/IntelliJ IDEA.app/Contents/lib/netty-resolver-4.1.52.Final.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/javac2.jar:/Applications/IntelliJ IDEA.app/Contents/lib/httpclient-4.5.12.jar:/Applications/IntelliJ IDEA.app/Contents/lib/plexus-component-annotations-1.7.1.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-resolver-spi-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/util.jar:/Applications/IntelliJ IDEA.app/Contents/lib/platform-api.jar:/Applications/IntelliJ IDEA.app/Contents/lib/qdox-2.0-M10.jar:/Applications/IntelliJ IDEA.app/Contents/lib/asm-all-9.0.jar:/Applications/IntelliJ IDEA.app/Contents/lib/commons-lang3-3.10.jar:/Applications/IntelliJ IDEA.app/Contents/lib/jna.jar:/Applications/IntelliJ IDEA.app/Contents/lib/trove4j.jar:/Applications/IntelliJ IDEA.app/Contents/lib/nanoxml-2.2.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-resolver-api-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/maven-resolver-connector-basic-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/plexus-utils-3.2.0.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-resolver-impl-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/jdom.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-resolver-util-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/protobuf-java-3.13.0.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/aether-dependency-resolver.jar:/Applications/IntelliJ IDEA.app/Contents/lib/netty-buffer-4.1.52.Final.jar:/Applications/IntelliJ IDEA.app/Contents/lib/forms-1.1-preview.jar:/Applications/IntelliJ IDEA.app/Contents/lib/annotations.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-resolver-provider-3.6.1.jar:/Applications/IntelliJ IDEA.app/Contents/lib/resources_en.jar:/Applications/IntelliJ IDEA.app/Contents/lib/netty-transport-4.1.52.Final.jar:/Applications/IntelliJ IDEA.app/Contents/lib/netty-codec-4.1.52.Final.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-model-3.6.1.jar:/Applications/IntelliJ IDEA.app/Contents/lib/guava-29.0-jre.jar:/Applications/IntelliJ IDEA.app/Contents/lib/intellij-deps-fastutil-8.4.1-4.jar:/Applications/IntelliJ IDEA.app/Contents/lib/lz4-java-1.7.1.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/jps-javac-extension-1.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/jps-builders-6.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/maven-resolver-transport-file-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/jna-platform.jar:/Applications/IntelliJ IDEA.app/Contents/lib/oro-2.0.8.jar:/Applications/IntelliJ IDEA.app/Contents/lib/plexus-interpolation-1.25.jar:/Applications/IntelliJ IDEA.app/Contents/lib/jps-model.jar:/Applications/IntelliJ IDEA.app/Contents/lib/gson-2.8.6.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-model-builder-3.6.1.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-builder-support-3.6.1.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/maven-resolver-transport-http-1.3.3.jar:/Applications/IntelliJ IDEA.app/Contents/lib/commons-logging-1.2.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/jps-builders.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-artifact-3.6.1.jar:/Applications/IntelliJ IDEA.app/Contents/lib/log4j.jar:/Applications/IntelliJ IDEA.app/Contents/lib/httpcore-4.4.13.jar:/Applications/IntelliJ IDEA.app/Contents/lib/slf4j-api-1.7.25.jar:/Applications/IntelliJ IDEA.app/Contents/lib/idea_rt.jar:/Applications/IntelliJ IDEA.app/Contents/lib/maven-repository-metadata-3.6.1.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/uiDesigner/lib/jps/java-guiForms-jps.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/Groovy/lib/groovy-jps.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/Groovy/lib/groovy-constants-rt.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/maven/lib/maven-jps.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/gradle-java/lib/gradle-jps.jar:/Applications/IntelliJ IDEA.app/Contents/plugins/devkit/lib/devkit-jps.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/jps/kotlin-jps-plugin.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-stdlib.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-stdlib-jdk7.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-stdlib-jdk8.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-reflect.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-plugin.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-jps-common.jar:/Users/parrt/Library/Application Support/JetBrains/IntelliJIdea2020.3/plugins/Kotlin/lib/kotlin-common.jar" org.jetbrains.jps.cmdline.BuildMain 127.0.0.1 50242 d741d84a-64e8-4ad3-9dcf-02d5f1dd1455 /Users/parrt/Library/Caches/JetBrains/IntelliJIdea2020.3/compile-server
Build process started. Classpath: /Applications/IntelliJ IDEA.app/Contents/plugins/java/lib/jps-launcher.jarSLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.intellij.util.ReflectionUtil (file:/Applications/IntelliJ%20IDEA.app/Contents/lib/util.jar) to method java.util.ResourceBundle.setParent(java.util.ResourceBundle)
WARNING: Please consider reporting this to the maintainers of com.intellij.util.ReflectionUtil
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future releaseJAVAC_PROCESS[STDERR]: java.lang.RuntimeException: java.lang.ClassCastException: com.sun.proxy.$Proxy11 cannot be cast to com.sun.tools.javac.processing.JavacProcessingEnvironment
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.main.Main.compile(Main.java:469)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:132)
JAVAC_PROCESS[STDERR]: 	at org.jetbrains.jps.javac.JavacMain.compile(JavacMain.java:231)
JAVAC_PROCESS[STDERR]: 	at org.jetbrains.jps.javac.ExternalJavacProcess.compile(ExternalJavacProcess.java:190)
JAVAC_PROCESS[STDERR]: 	at org.jetbrains.jps.javac.ExternalJavacProcess.access$300(ExternalJavacProcess.java:32)
JAVAC_PROCESS[STDERR]: 	at org.jetbrains.jps.javac.ExternalJavacProcess$CompilationRequestsHandler$1.run(ExternalJavacProcess.java:262)
JAVAC_PROCESS[STDERR]: 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
JAVAC_PROCESS[STDERR]: 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
JAVAC_PROCESS[STDERR]: 	at java.lang.Thread.run(Thread.java:745)
JAVAC_PROCESS[STDERR]: Caused by: java.lang.ClassCastException: com.sun.proxy.$Proxy11 cannot be cast to com.sun.tools.javac.processing.JavacProcessingEnvironment
JAVAC_PROCESS[STDERR]: 	at org.antlr.v4.test.runtime.CommentHasStringValueProcessor.init(CommentHasStringValueProcessor.java:49)
JAVAC_PROCESS[STDERR]: 	at org.jetbrains.jps.javac.APIWrappers$ProcessorWrapper.init(APIWrappers.java:105)
JAVAC_PROCESS[STDERR]: 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
JAVAC_PROCESS[STDERR]: 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
JAVAC_PROCESS[STDERR]: 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
JAVAC_PROCESS[STDERR]: 	at java.lang.reflect.Method.invoke(Method.java:606)
JAVAC_PROCESS[STDERR]: 	at org.jetbrains.jps.javac.APIWrappers$1.invoke(APIWrappers.java:213)
JAVAC_PROCESS[STDERR]: 	at com.sun.proxy.$Proxy10.init(Unknown Source)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.processing.JavacProcessingEnvironment$ProcessorState.<init>(JavacProcessingEnvironment.java:517)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.processing.JavacProcessingEnvironment$DiscoveredProcessors$ProcessorStateIterator.next(JavacProcessingEnvironment.java:614)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:707)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1700(JavacProcessingEnvironment.java:97)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1029)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1163)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1108)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:824)
JAVAC_PROCESS[STDERR]: 	at com.sun.tools.javac.main.Main.compile(Main.java:439)
JAVAC_PROCESS[STDERR]: 	... 8 more
```

I had to manually kill the Java process to get the development environment to stop. note that it is using java from the development environment and not java 7 as I say in the project or what maven says.

```bash
$ '/Applications/IntelliJ IDEA.app/Contents/jbr/Contents/Home/bin/java' --version
openjdk 11.0.9.1 2020-11-04
OpenJDK Runtime Environment JBR-11.0.9.1.11-1145.63-jcef (build 11.0.9.1+11-b1145.63)
OpenJDK 64-Bit Server VM JBR-11.0.9.1.11-1145.63-jcef (build 11.0.9.1+11-b1145.63, mixed mode)
```

**Reply by ericvergnaud** [reply]
> Hi,
> 
> I just spent a man/day trying to get a GitHub self-hosted runner to work on MacOS for Swift, to no end.
> It's able to fetch a commit and run all the builds but fails miserably at running tests (it runs 1 test file and then hangs forever).
> If I run the test script from the cmd line it works.
> So I suspect it has something to do with the GitHub runner getting confused by the child processes being created, a model which is core to our testing architecture.
> It might also be related to the GitHub runner never properly tested on MacOSX (as an example there is no predefined action to install maven...)
> If someone wants to try it out feel free...
> I will be looking at alternatives.

Ok after digging a bit further it seems the culprit is VirtualBox (not yet fully compatible with Big Sur).
Running the actions on Big Sur directly works!

So it seems a 100% free resolution path could be:
 - keep Wintel on AppVeyor 
 - move MacOS to Github actions self hosted
 - move Linux to Circle CI

**Comment by parrt** [comment]
That works to me if circle can give us enough free builds...

**Comment by prankkelkar** [comment]
Hi All. Is the migration complete? I dont see jobs running on Travis anymore  however `.travis.yml` file is still present.

**Reply by ericvergnaud** [reply]
Migration is complete, however Swift/Linux is still on Travis CI for technical reasons.
They only provide limited credit to OSS projects hence why you don't see many builds.

**Reply by prankkelkar** [reply]
Thanks for the information.
